pkg/storage/stores/shipper/downloads/table_manager.go

```diff
@@ -153,31 +153,27 @@ func (tm *TableManager) query(ctx context.Context, tableName string, queries []c
 }
 
 func (tm *TableManager) getOrCreateTable(tableName string) (*Table, error) {
-	// if table is already there, use it.
 	tm.tablesMtx.RLock()
+	defer tm.tablesMtx.RUnlock()
```

Comment:
I think we're fixing a deadlock and introducing a new one. We use `RLock` first as a lower-cost way to check if the table exists, but must `RUnlock` it after (the previous PR did this). We cannot `defer` it here because we need to `Lock` it later if the table wasn't found. I think we can introduce a 1 line fix:`defer tm.tablesMtx.Unlock()` which will ensure we release the write lock correctly in all cases it's acquired.

Reply:
Yeah, I missed the `Rlock` vs. `Lock`

The latest commit just defers `tm.tablesMtx.Unlock()` after it's created.